### PR TITLE
test: add some scheduler integration tests (1/)

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR adds some integration to the scheduler and its related components.

The added test suite verifies the behavior of the scheduler when it processed CRPs of the `PickFixed` placement type.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

N/A

### CODE REVIEW

File: cmd/hubagent/workload/setup.go
Comment (ryanzhang-oss) [inline cmd/hubagent/workload/setup.go:215]: is Workqueue a word?

File: pkg/scheduler/framework/framework_test.go
Comment (ryanzhang-oss) [inline pkg/scheduler/framework/framework_test.go:4543]: Are those additional changes needed after the CRD split? Will they disappear after rebase?

File: cmd/hubagent/workload/setup.go
Comment (michaelawyu) [reply cmd/hubagent/workload/setup.go:215]: Both spelling exist in the Kubernetes repo (and across the net); shall we have a preferred one? I guess `workqueue` is used because of the `workqueue` package 🤔

File: pkg/scheduler/framework/framework_test.go
Comment (michaelawyu) [reply pkg/scheduler/framework/framework_test.go:4543]: Yeah, they will disappear after rebasing. I am rebasing the PR now.

File: cmd/hubagent/workload/setup.go
Comment (zhiying-lin) [reply cmd/hubagent/workload/setup.go:215]: prefer WorkQueue as they are two separate words.

File: Makefile
Comment (zhiying-lin) [inline Makefile:135]: can we make it as a part of "test" target? it won't be missed when others run make test command to validate the changes.

File: test/scheduler/actuals_test.go
Comment (zhiying-lin) [inline test/scheduler/actuals_test.go:124]:
```suggestion
                options := []cmp.Option{
				ignoreResourceBindingTypeMetaField,
			        ignoreObjectMetaNameField,
			        ignoreObjectMetaAutoGeneratedFields,
			        ignoreClusterDecisionReasonField,
			        cmpopts.SortSlices(lessFuncBinding),
			}
		if diff := cmp.Diff(scheduled, wantScheduled, options...),
```
refactor it as a common variable? it is used in multiple places

File: test/scheduler/suite_test.go
Comment (zhiying-lin) [inline test/scheduler/suite_test.go:42]:
```suggestion
	"go.goms.io/fleet/pkg/scheduler"
	"go.goms.io/fleet/pkg/scheduler/clustereligibilitychecker"
	"go.goms.io/fleet/pkg/scheduler/framework"
	"go.goms.io/fleet/pkg/scheduler/framework/plugins/clusteraffinity"
	"go.goms.io/fleet/pkg/scheduler/framework/plugins/clustereligibility"
	"go.goms.io/fleet/pkg/scheduler/framework/plugins/sameplacementaffinity"
	"go.goms.io/fleet/pkg/scheduler/framework/plugins/topologyspreadconstraints"
	"go.goms.io/fleet/pkg/scheduler/queue"
	"go.goms.io/fleet/pkg/scheduler/watchers/clusterresourceplacement"
	"go.goms.io/fleet/pkg/scheduler/watchers/clusterschedulingpolicysnapshot"
	"go.goms.io/fleet/pkg/scheduler/watchers/membercluster"
```
missed this before, renaming is not necessary as there is no conflict.

& redundant empty line

File: test/scheduler/pickfixed_integration_test.go
Comment (ryanzhang-oss) [inline test/scheduler/pickfixed_integration_test.go:28]: is there any reason those clusters need to be selected? Are those just random?

File: test/scheduler/actuals_test.go
Comment (ryanzhang-oss) [inline test/scheduler/actuals_test.go:365]:
```suggestion
		// Verify that cluster decisions are populated correctly.
```

File: test/scheduler/actuals_test.go
Comment (ryanzhang-oss) [inline test/scheduler/actuals_test.go:366]:
```suggestion
		wantClusterDecisions := []fleetv1beta1.ClusterDecision{}
```

File: test/scheduler/pickfixed_integration_test.go
Comment (ryanzhang-oss) [inline test/scheduler/pickfixed_integration_test.go:134]: why reuse the previous test?

**Comment by ryanzhang-oss** [comment]
I ran the IT a few times locally but one test failed once.


Summarizing 1 Failure:
  [FAIL] scheduling CRPs of the PickFixed placement type create a CRP with some valid target clusters [It] should add scheduler cleanup finalizer to the CRP
  /Users/ryanzhang/workspace/github/fleet/test/scheduler/pickfixed_integration_test.go:82

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:103]: Done. It has been renamed using the pattern `cluster-[INDEX]-[REGION]-[ENV]`, e.g., `cluster-1-east-prod`.

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:214]: Done.

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:11]: Hi Ryan! It was put into its own separate folder mostly for 2 reasons:

1) The integration tests cover not only the scheduler itself but its related components (traditionally the integration tests we included in one package cares only for that package and not beyond).
2) To enable parallelization (the scheduler package has unit tests, which is run by `go test`; if we place them together, Ginkgo tests will run in a single thread as well).

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:250]: Done.

File: test/scheduler/actuals_test.go
Comment (michaelawyu) [reply test/scheduler/actuals_test.go:365]: Done for other occurrences as well.

File: test/scheduler/actuals_test.go
Comment (michaelawyu) [reply test/scheduler/actuals_test.go:366]: Done.

**Comment by michaelawyu** [comment]
> I don't see detailed errors in the output. Can one see the server side logs on the command line output?

Hi Ryan! I have run the test for multiple times and could not reproduce the error. Regarding the log outputs, it's a verbosity level (klog uses default 0); I have updated the setup to use 5 instead.

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:42]: I don't particularly have a strong opinion on this one, though I do want to add that it's quite common in practice to rename K8s package for simplicity reasons, i.e., `controller-runtime` -> `ctrl`

Comment (michaelawyu) [reply test/scheduler/suite_test.go:42]: And the scheduler/framework rename is to avoid name collisions in the code.

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go:357]: Done.

File: test/scheduler/actuals_test.go
Comment (michaelawyu) [reply test/scheduler/actuals_test.go:124]: Done.

File: test/scheduler/suite_test.go
Comment (michaelawyu) [reply test/scheduler/suite_test.go]: Done.

File: test/scheduler/suite_test.go
Comment (ryanzhang-oss) [inline test/scheduler/suite_test.go:64]: the real value of the cluster is not visible in the test cases. It's the name of the variable that really matters. In test case, I still see pick cluster1,2,3 instead verify cluster-1-east, cluster-2-east, cluster-3-east in the test case.

**Review by ryanzhang-oss — APPROVED** [review]

**Comment by michaelawyu** [comment]
Merged this PR to unblock the progress. Has acquired approval before.

If there's any concern, please let me know.

**Pull request merged**
